src/DFMM.sol

```diff
@@ -278,33 +281,43 @@ contract DFMM is IDFMM {
     // Internals
 
     /**
-     * @dev Transfers `amount` of `token` from the sender to the contract. Note
-     * that if ETH is present in the contract, it will be wrapped to WETH. Any
-     * excess of ETH will be sent back to the sender.
-     * @param token Address of the token to transfer.
-     * @param amount Amount to transfer expressed in WAD.
+     * @dev Transfers `amounts` of `tokens` from the sender to the contract. Note
+     * that if any ETH is present in the contract, it will be wrapped to WETH and
+     * used if sufficient. Any excess of ETH will be sent back to the sender.
+     * @param tokens An array of token addresses to transfer.
+     * @param amounts An array of amounts to transfer expressed in WAD.
      */
-    function _transferFrom(address token, uint256 amount) internal {
-        if (address(this).balance >= amount) {
-            WETH(payable(weth)).deposit{ value: amount }();
+    function _transferFrom(
+        address[] memory tokens,
+        uint256[] memory amounts
+    ) internal {
+        uint256 length = tokens.length;
+
+        for (uint256 i = 0; i < length; i++) {
+            address token = tokens[i];
+            uint256 amount = amounts[i];
 
-            if (address(this).balance > 0) {
-                SafeTransferLib.safeTransferETH(
-                    msg.sender, address(this).balance
-                );
-            }
-        } else {
             uint256 downscaledAmount =
                 downscaleUp(amount, computeScalingFactor(token));
             uint256 preBalance = ERC20(token).balanceOf(address(this));
-            SafeTransferLib.safeTransferFrom(
-                ERC20(token), msg.sender, address(this), downscaledAmount
-            );
+
+            if (token == weth && address(this).balance >= amount) {
+                WETH(payable(weth)).deposit{ value: amount }();
+            } else {
```

Comment:
It's worth noting that an attacker can still "pay on behalf" of a victim by directly sending an ETH balance of `amount` to the contract, causing this branch to be taken instead of the one below.

Perhaps this might cause problems for certain external integrations that expect their WETH balance to decrease after calling `init()`, `allocate()` or `swap()`. Additionally, it would leave a dangling WETH approval from the calling contract to the `DFMM` contract.

Is there any reason we don't use `msg.value` here instead of `address(this).balance`?

```diff
- if (token == weth && address(this).balance >= amount) {
+ if (token == weth && msg.value >= amount) {
      WETH(payable(weth)).deposit{ value: amount }();
  } else {
```

This would prevent any attacker from interfering with a victim's calls by sending ETH to the contract.

Reply:
Same comment as above, in the end we might want to get rid of raw ETH completely.

Reply:
Fix pr: https://github.com/primitivefinance/DFMM/pull/122